### DISCUSSION TITLE
Unable to find libclang (linux/x86_64)

### DISCUSSION THREAD

**Discussion by ods** [messense/maturin-action#152] [Q&A] [answered]
There is a CLI tool written in rust ([squawk](https://github.com/sbdchd/squawk) and I'd like to create a python package for it (`bindings = "bin"`). The Github Action workflow for original package just [installs the rust toolchain and runs `cargo build`](https://github.com/sbdchd/squawk/blob/master/.github/workflows/rust.yml#L27-L36), nothing special. Running `PyO3/maturin-action@v1` jobs for my package works fine on macos, but fails on linux with the following error:
```
error: failed to run custom build command for `libpg_query-sys v0.2.1`
Caused by:
  process didn't exit successfully: `/home/runner/work/squawk-cli-python-package/squawk-cli-python-package/origin/target/release/build/libpg_query-sys-8fad67c095f7e4fb/build-script-build` (exit status: 101)
  --- stdout
  cargo:rustc-link-lib=static=pg_query
  cargo:rustc-link-search=native=/home/runner/work/squawk-cli-python-package/squawk-cli-python-package/origin/target/x86_64-unknown-linux-gnu/release/build/libpg_query-sys-a8de555d92b0457e/out/libpg_query
  --- stderr
  thread 'main' panicked at 'Unable to find libclang: "couldn't find any valid shared libraries matching: ['libclang.so', 'libclang-*.so', 'libclang.so.*', 'libclang-*.so.*'], set the `LIBCLANG_PATH` environment variable to a path where one of these files can be found (invalid: [])"', /root/.cargo/registry/src/github.com-1ecc6299db9ec823/bindgen-0.60.1/src/lib.rs:2172:31
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
warning: build failed, waiting for other jobs to finish...
💥 maturin failed
  Caused by: Failed to build a native library through cargo
  Caused by: Cargo build finished with "exit status: 101": `"cargo" "rustc" "--release" "--target" "x86_64-unknown-linux-gnu" "--manifest-path" "/home/runner/work/squawk-cli-python-package/squawk-cli-python-package/origin/cli/Cargo.toml" "--message-format" "json" "--bin" "squawk" "--" "-C" "link-arg=-s"
```

While debugging I've checked that libclang.so is installed. Also I've tried to LIBCLANG_PATH to the `/usr/lib/llvm-14/lib` (as returned by `dpkg -S libclang.so` command) and added `docker-options: -e LIBCLANG_PATH` to insure this var is passed to the container. But the error remains the same. 

Any other idea what can be wrong?

**Comment by messense** [comment, marked as answer]
`maturin-action` builds linux wheels in docker containers, so you can't possibly pass host libclang to it.

`quay.io/pypa/manylinux2014_x86_64` doesn't install `libclang` I think.

**Reply by ods** [reply]
Thank you for the response. So, `actions-rs/toolchain` installs everything (including `libclang`) into top level container, while `PyO3/maturin-action` builds everything in a separate container (docker in docker). That's why my approach doesn't work, right?

Then I have to options:
- find a way to install additional packages into the container, used by `PyO3/maturin-action`,
- instead of using `PyO3/maturin-action` install maturin into the top-level container and write all the needed commands to build directly in pipeline file.

With later everything is clear so far, but it requires a bunch of work. Any ideas how to proceed with the first option?

**Reply by messense** [reply]
> So, `actions-rs/toolchain` installs everything (including `libclang`) into top level container

No, `actions-rs/toolchain` installs everything to github runner unless you are using the github actions `container:` option or using `cross`

> while `PyO3/maturin-action` builds everything in a separate container (docker in docker)

No, there are `docker in docker` involved, it's simply a `docker run` command on github runner

> With later everything is clear so far, but it requires a bunch of work. Any ideas how to proceed with the first option?

Currently no, but you can build a docker image with clang installed and pass it to `container` option of `maturin-action`.

**Reply by ods** [reply]
Thanks!

**Reply by messense** [reply]
Add `before-script-linux` in https://github.com/PyO3/maturin-action/pull/157 to make it easier.
